File: tests/_stall_worker.py
"""Adversarial worker variant for the temporal-oracle tests (Binding 4.4).

Behaves exactly like ``voss.worker`` (proposes over the authenticated
channel, never executes) but can be made to **stall** past the runtime's
declared response ceiling (``VOSS_STALL_SECONDS``), or to **flood** each
proposal line with an oversized volume (``VOSS_BIG_VOLUME=1``).  Either makes
the runtime's temporal oracle register anomalies and engage containment.
"""
from __future__ import annotations

import os
import sys
import time

from voss.canonical import new_id
from voss.chan import ChannelSession, read_bootstrap_file, unlink_quiet


class AdversarialModel:
    """Deterministic proposals; may stall or bloat on request."""

    @staticmethod
    def _env(action: str, session_id: str, principal: str,
             payload: dict) -> dict:
        return {
            "version": "1",
            "request_id": new_id("req-"),
            "session_id": session_id,
            "principal": principal,
            "action": action,
            "resource": {"path": "notes.txt"},
            "payload": payload,
            "constraints": {},
        }

    def propose(self, prompt: str, session_id: str, principal: str) -> list:
        stall = float(os.environ.get("VOSS_STALL_SECONDS", "0") or 0)
        time.sleep(stall)
        payload: dict = {}
        if os.environ.get("VOSS_BIG_VOLUME") == "1":
            # ~40 KiB line: under the 64 KiB channel cap but over the
            # temporal oracle's volume cap.
            payload = {"noise": "x" * 40000}
        return [self._env("workspace.read", session_id, principal, payload)]


def main() -> int:
    bootstrap_path = os.environ.get("VOSS_CHANNEL_BOOTSTRAP")
    if not bootstrap_path:
        print("error: no channel bootstrap; refusing unauthenticated mode",
              file=sys.stderr)
        return 2
    try:
        key, sid = read_bootstrap_file(bootstrap_path)
        unlink_quiet(bootstrap_path)
    except Exception as exc:
        print(f"error: cannot load channel bootstrap: {exc}",
              file=sys.stderr)
        return 2
    if not key or not sid:
        return 2

    session = ChannelSession(key, sid, "adapter")
    model = AdversarialModel()
    print(session.send("hello", {"adapter": "voss.adversarial-stall",
                                 "ready": True}), flush=True)
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg_type, msg = session.receive(line)
        except Exception as exc:
            print(f"error: host channel violation: {exc}", file=sys.stderr)
            return 3
        if msg_type == "prompt":
            envelopes = model.propose(
                str(msg.get("prompt", "")),
                str(msg.get("session_id", "")),
                str(msg.get("principal", "")),
            )
            if envelopes:
                print(session.send("proposal", {"envelopes": envelopes}),
                      flush=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/src/bin/stall_worker.rs
//! Adversarial worker for the temporal oracle.
//!
//! Same authenticated channel as the real adapter, and it never executes.
//! `VOSS_STALL_SECONDS` delays each proposal. `VOSS_BIG_VOLUME=1` pads the
//! proposal past the 16 KiB exchange cap while staying under the channel
//! line limit.

use std::io::{self, BufRead, Write};
use std::path::Path;
use std::thread;
use std::time::Duration;

use voss::canonical::{Json, new_id};
use voss::chan::{self, ChannelSession, MAX_LINE};

fn main() {
    std::process::exit(run());
}

fn run() -> i32 {
    let path = match std::env::var("VOSS_CHANNEL_BOOTSTRAP") {
        Ok(path) if !path.is_empty() => path,
        _ => {
            eprintln!("error: no channel bootstrap; refusing unauthenticated mode");
            return 2;
        }
    };
    let (key, sid) = match chan::consume_bootstrap(Path::new(&path)) {
        Ok(value) => value,
        Err(error) => {
            eprintln!("error: {}", error.code());
            return 2;
        }
    };
    let mut session = match ChannelSession::new(key, sid, "adapter") {
        Ok(session) => session,
        Err(error) => {
            eprintln!("error: {}", error.code());
            return 2;
        }
    };
    let hello = Json::object([
        ("adapter", Json::string("voss.adversarial-stall")),
        ("ready", Json::Bool(true)),
    ]);
    let Ok(hello_line) = session.send("hello", &hello) else {
        eprintln!("error: denied_channel_auth");
        return 3;
    };
    println!("{hello_line}");
    let _ = io::stdout().flush();

    let stdin = io::stdin();
    let mut reader = stdin.lock();
    loop {
        let line = match read_capped(&mut reader) {
            ReadEvent::Line(line) => line,
            ReadEvent::Empty => continue,
            ReadEvent::Eof => break,
            ReadEvent::Oversize | ReadEvent::Failed => {
                eprintln!("error: denied_channel_oversize");
                return 3;
            }
        };
        let (msg_type, msg) = match session.receive(&line) {
            Ok(value) => value,
            Err(error) => {
                eprintln!("error: {}", error.code());
                return 3;
            }
        };
        if msg_type == "hello_ok" {
            continue;
        }
        if msg_type != "prompt" {
            continue;
        }
        let stall = stall_seconds();
        if stall > 0.0 {
            thread::sleep(Duration::from_secs_f64(stall));
        }
        let session_id = msg.get("session_id").and_then(Json::as_str).unwrap_or("");
        let principal = msg.get("principal").and_then(Json::as_str).unwrap_or("");
        let payload = if std::env::var("VOSS_BIG_VOLUME").ok().as_deref() == Some("1") {
            Json::object([("noise", Json::string("x".repeat(40_000)))])
        } else {
            Json::empty_object()
        };
        let envelopes = vec![Json::object([
            ("action", Json::string("workspace.read")),
            ("constraints", Json::empty_object()),
            ("payload", payload),
            ("principal", Json::string(principal)),
            ("request_id", Json::string(new_id("req-"))),
            ("resource", Json::object([("path", Json::string("notes.txt"))])),
            ("session_id", Json::string(session_id)),
            ("version", Json::string("1")),
        ])];
        let proposal = Json::object([("envelopes", Json::Array(envelopes))]);
        let Ok(wire) = session.send("proposal", &proposal) else {
            eprintln!("error: denied_channel_auth");
            return 3;
        };
        println!("{wire}");
        let _ = io::stdout().flush();
    }
    0
}

fn stall_seconds() -> f64 {
    std::env::var("VOSS_STALL_SECONDS")
        .ok()
        .filter(|value| !value.is_empty())
        .and_then(|value| value.parse::<f64>().ok())
        .filter(|value| value.is_finite() && *value > 0.0)
        .unwrap_or(0.0)
}

enum ReadEvent {
    Line(String),
    Empty,
    Eof,
    Oversize,
    Failed,
}

fn read_capped(reader: &mut impl BufRead) -> ReadEvent {
    let mut raw = Vec::new();
    let mut byte = [0u8; 1];
    loop {
        match reader.read(&mut byte) {
            Ok(0) => {
                if raw.is_empty() {
                    return ReadEvent::Eof;
                }
                break;
            }
            Ok(_) => {
                raw.push(byte[0]);
                if raw.len() > MAX_LINE {
                    return ReadEvent::Oversize;
                }
                if byte[0] == b'\n' {
                    break;
                }
            }
            Err(_) => return ReadEvent::Failed,
        }
    }
    while raw.last() == Some(&b'\n') || raw.last() == Some(&b'\r') {
        raw.pop();
    }
    match String::from_utf8(raw) {
        Ok(line) if line.is_empty() => ReadEvent::Empty,
        Ok(line) => ReadEvent::Line(line),
        Err(_) => ReadEvent::Failed,
    }
}
